pkg/crd/crdvalidations.go

```diff
@@ -0,0 +1,412 @@
+		OpenAPIV3Schema: &packageSchema,
+	}
+)
+
+// Children of Package crd schema
+var (
+	archiveSchemaProps = map[string]apiextensionsv1beta1.JSONSchemaProps{
+		"type": {
+			Type:        "string",
+			Description: "Type defines how the package is specified: literal or url.",
+		},
+		"literal": {
+			Type:        "string",
+			Format:      "byte",
+			Description: "Literal contents of the package.",
+		},
+		"url": {
+			Type:        "string",
+			Description: "URL references a package.",
+		},
+		"checksum": checksumSchema,
+	}
+	archiveSchema = apiextensionsv1beta1.JSONSchemaProps{
+		Type:        "object",
+		Description: "Package contains or references a collection of source or binary files.",
+		Properties:  archiveSchemaProps,
+	}
+)
+
+var (
+	checksumSchemaProps = map[string]apiextensionsv1beta1.JSONSchemaProps{
+		"type": {
+			Type:        "string",
+			Description: "ChecksumType specifies the checksum algorithm, such as sha256, used for a checksum.",
+		},
+		"sum": {
+			Type:        "string",
+			Description: " Sum is hex encoded chechsum value.",
+		},
+	}
+	checksumSchema = apiextensionsv1beta1.JSONSchemaProps{
+		Type:        "object",
+		Description: "Checksum of package contents when the contents are stored outside the Package struct. Type is the checksum algorithm;  sha256 is the only currently supported one. Sum is hex  encoded.",
+		Properties:  checksumSchemaProps,
+	}
+)
+
+// Children of Function crd schema
+var (
+	environmentReferenceSchemaProps = map[string]apiextensionsv1beta1.JSONSchemaProps{
+		"namespace": {
+			Type:        "string",
+			Description: "Namespace for corresponding Environment",
+		},
+		"name": {
+			Type:        "string",
+			Description: "Name of the Environment to use",
+		},
+	}
+	environmentReferenceSchema = apiextensionsv1beta1.JSONSchemaProps{
+		Type:        "object",
+		Description: "Reference to Fission Environment type custom resource.",
+		Properties:  environmentReferenceSchemaProps,
+	}
+)
+
+var (
+	packageRefSchemaProps = map[string]apiextensionsv1beta1.JSONSchemaProps{
+		"namespace": {
+			Type:        "string",
+			Description: "Namespace for corresponding Package",
+		},
+		"name": {
+			Type:        "string",
+			Description: "Name of the Package to use",
+		},
+		"resourceversion": {
+			Type:        "string",
+			Description: "Including resource version in the reference forces the function to be updated on package update, making it possible to cache the function based on its metadata.",
+		},
+	}
+	functionPackageRefSchemaProps = map[string]apiextensionsv1beta1.JSONSchemaProps{
+		"packageref": {
+			Type:        "object",
+			Description: "Package Reference",
+			Properties:  packageRefSchemaProps,
+		},
+		"functionName": {
+			Type:        "string",
+			Description: "FunctionName specifies a specific function within the package. This allows functions to share packages, by having different functions within the same package.",
```

Comment:
Description: "FunctionName specifies a specific function within the package using the path and specific function and varies based on language/environment",